validphys2/src/validphys/chi2grids.py

```diff
@@ -4,109 +4,71 @@
     """
-
-    #TODO: Everythning about this function is horrible. We need to rewrite
-    #experiments.cc from scratch.
-
-    #TODO: Do this somewhere else
-    RandomGenerator.InitRNG(0,0)
-    if t0set is not None:
-        lt0 = t0set.load_t0()
-    pdfname = pdf.name
-    datas = []
-
-    #No need to save these in the cache, so we call __wrapped__
-    original_experiments = [e.load.__wrapped__(e) for e in experiments]
-    sqrtcovmat_table = []
-    log.debug("Generating dataset covmats")
-    for exp in original_experiments:
-        if t0set is not None:
-            exp.SetT0(lt0)
-        #The covariance matrices are currently very expensive to recompute.
-        #Store them after computing T0
-        sqrtcovmat_table.append([ds.get_sqrtcovmat() for ds in exp.DataSets()])
-
-    for lhapdf_index, nnfit_index in enumerate(fitted_replica_indexes, 1):
-
-        flutuated_experiments = pseudodata(original_experiments, dataseed, nnfit_index)
-        lpdf = single_replica(pdfname, lhapdf_index)
-        for expspec, exp, mats in zip(experiments, flutuated_experiments, sqrtcovmat_table):
-            #We need to manage the memory
-            exp.thisown = True
-
-            th = ThPredictionsResult.from_convolution(pdf, expspec,
-                loaded_data=exp, loaded_pdf=lpdf)
-
-
-            results = DataResult(exp, exp.get_covmat(), exp.get_sqrtcovmat()), th
-            #The experiment already has T0. No need to set it again.
-            #TODO: This is a hack. Get rid of this.
-            chi2 = chi2_breakdown_by_dataset(results, exp, t0set=None,
-                                             datasets_sqrtcovmat=mats)
-
-            for i, (dsname,(value, ndata)) in enumerate(chi2.items()):
-                data = PseudoReplicaExpChi2Data(
-                    nnfit_index=nnfit_index,
-                    experiment=expspec.name,
-                    #We set the i so that the sort order is maintaned here.
-                    dataset = (i, dsname),
-                    ndata = ndata,
-                    chi2=value
-                    )
-                datas.append(data)
-
-    df =  pd.DataFrame(datas, columns=PseudoReplicaExpChi2Data._fields)
-    df.set_index(['experiment', 'dataset', 'ndata' , 'nnfit_index'], inplace=True)
+    # Stack the replica pseudodata to have the prediction shape
+    r_data = np.stack(fitted_make_replicas, axis=1)
+
+    # Drop data central and theory central which is not useful here
+    r_prediction = group_result_table_no_table.drop(columns=["data_central", "theory_central"])
+
+    # Now compute the chi2 in a per-group basis
+    diff = r_prediction - r_data
+    group_level = r_prediction.index.get_level_values("group")
+
+    # Save the results in a dataframe similar (but not equal) to the old one
+    df_output = []
+    for group in group_level.unique():
+        group_diff = diff.loc[group_level == group]
+        its_covmat = groups_sqrtcovmat[group_level == group][group]
```

Comment:
How come you don't need `[group_level == group]` for the second slice? Also would the syntax not be `[group_level == group, group]`?

Reply:
Oh I see what's going on here, never mind, multi index is annoying

Reply:
Maybe `groups_sqrtcovmat.loc[group][group]` will be a bit more readable what do you think?

Reply:
No, the results are not equivalent so you need afterwards a line to convert while this one works ootb.